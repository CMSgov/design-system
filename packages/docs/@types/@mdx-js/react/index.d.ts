declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
