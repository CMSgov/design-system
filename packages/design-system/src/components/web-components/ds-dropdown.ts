import { define } from 'preactement';
import Dropdown from '../Dropdown/Dropdown';

const attributes = [
  'auto-focus',
  'class-name',
  'disabled',
  'error-placement',
  'field-class-name',
  'label',
  'label-class-name',
  'label-id',
  'name',
  'role',
  'size',
  'value',
  'default-value',
] as const;

declare global {
  namespace JSX {
    interface IntrinsicElements {
      'ds-dropdown': React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement> & {
          [K in typeof attributes[number]]?: string;
        },
        HTMLElement
      >;
    }
  }
}

define('ds-dropdown', () => Dropdown, { attributes });
