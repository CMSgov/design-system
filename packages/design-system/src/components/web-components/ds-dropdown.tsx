import React from 'react';
import { define } from './preactement/define';
import { Dropdown, DropdownProps } from '../Dropdown';

const attributes = [
  'auto-focus',
  'class-name',
  'disabled',
  'error-message',
  'error-placement',
  'field-class-name',
  'options',
  'label',
  'label-class-name',
  'label-id',
  'name',
  'requirements-label',
  'role',
  'root-id',
  'size',
  'value',
  'default-value',
] as const;

/* eslint-disable @typescript-eslint/no-namespace */
declare global {
  namespace JSX {
    interface IntrinsicElements {
      'ds-dropdown': React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement> & {
          [K in (typeof attributes)[number]]?: string;
        },
        HTMLElement
      >;
    }
  }
}
/* eslint-enable */

interface WrapperProps extends Omit<DropdownProps, 'options'> {
  options?: string | DropdownProps['options'];
  rootId?: string;
}

const Wrapper = ({ children, options, rootId, ...otherProps }: WrapperProps) => (
  <Dropdown
    {...otherProps}
    options={typeof options === 'string' ? JSON.parse(options) : options}
    id={rootId}
  >
    {options ? undefined : children}
  </Dropdown>
);

define('ds-dropdown', () => Wrapper, { attributes } as any);
