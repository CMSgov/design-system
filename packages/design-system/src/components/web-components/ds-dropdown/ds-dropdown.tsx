import type * as React from 'react';
import { define } from '../preactement/define';
import { Dropdown, DropdownProps } from '../../Dropdown';
import { parseBooleanAttr } from '../wrapperUtils';
import { formAttrs } from '../shared-attributes/form';

const attributes = [
  // Using the lowercase HTML attribute name rather than `auto-focus` so it's
  // more natural. There's no reason for us to worry about name collisions.
  'autofocus',
  'aria-disabled',
  'class-name',
  'default-value',
  'disabled',
  'field-class-name',
  'inversed',
  'name',
  'options',
  'root-id',
  'size',
  'value',
  ...formAttrs,
] as const;

/* eslint-disable @typescript-eslint/no-namespace */
declare global {
  namespace JSX {
    interface IntrinsicElements {
      'ds-dropdown': React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement> & {
          [K in (typeof attributes)[number]]?: string;
        },
        HTMLElement
      >;
    }
  }
}
/* eslint-enable */

interface WrapperProps
  extends Omit<DropdownProps, 'options' | 'autoFocus' | 'disabled' | 'ariaDisabled' | 'inversed'> {
  autofocus?: string;
  ariaDisabled?: string;
  disabled?: string;
  inversed?: string;
  options?: string | DropdownProps['options'];
  rootId?: string;
}

const Wrapper = ({ children, options, rootId, ...otherProps }: WrapperProps) => (
  <Dropdown
    {...otherProps}
    autoFocus={parseBooleanAttr(otherProps.autofocus)}
    disabled={parseBooleanAttr(otherProps.disabled)}
    aria-disabled={parseBooleanAttr(otherProps.ariaDisabled)}
    options={typeof options === 'string' ? JSON.parse(options) : options}
    id={rootId}
    inversed={parseBooleanAttr(otherProps.inversed)}
  >
    {options ? undefined : children}
  </Dropdown>
);

define('ds-dropdown', () => Wrapper, { attributes, events: ['onChange', 'onBlur'] });
