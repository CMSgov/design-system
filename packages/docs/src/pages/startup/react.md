---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
