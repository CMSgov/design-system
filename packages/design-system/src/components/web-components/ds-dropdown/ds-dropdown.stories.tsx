import { useEffect } from 'react';
import WebComponentDocTemplate from '../../../../../../.storybook/docs/WebComponentDocTemplate.mdx';
import { action } from '@storybook/addon-actions';
import './ds-dropdown';
import { webComponentDecorator } from '../storybook';

const options = [
  { label: 'Confederated Tribes and Bands of the Yakama Nation', value: '1' },
  { label: 'Cowlitz Indian Tribe', value: '4' },
  { label: 'Lummi Tribe of the Lummi Reservation', value: '7' },
];

export default {
  title: 'Web Components/ds-dropdown',
  argTypes: {
    autofocus: {
      description: 'Sets the focus on the dropdown when it is first added to the document.',
      control: 'boolean',
    },
    children: { control: false },
    'class-name': {
      description: 'Additional classes to be added to the root button element.',
      control: 'text',
    },
    'default-value': {
      description:
        'Sets the initial selected state. Using this attribute instead of `value` means that the select will maintain its own internal selection state.',
      control: 'text',
    },
    disabled: { description: 'Disables the entire field.', control: 'boolean' },
    'error-message': {
      description: 'Enable the error state by providing an error message',
      control: 'text',
    },
    'error-id': {
      description:
        'The ID of the error message applied to this field. If none is provided, the id will be derived from the `root-id` attribute.',
      control: 'text',
    },
    'error-placement': {
      description: 'Location of the error message relative to the field input',
      options: [undefined, 'top', 'bottom'],
      control: { type: 'radio' },
    },
    'field-class-name': {
      description: 'Additional classes to be added to the dropdown button element',
      control: 'text',
    },
    hint: {
      description: 'Hint text or HTML',
      control: 'text',
    },
    'hint-class-name': {
      control: 'text',
      description: 'Custom CSS class name(s) for the hint element',
    },
    'hint-id': {
      description: 'The ID of the hint element',
      control: 'text',
    },
    inversed: {
      description: 'Set to "true" to apply the "inverse" theme',
      control: 'boolean',
    },
    label: {
      description: 'Label text or HTML.',
      control: 'text',
    },
    'label-class-name': {
      description: 'Additional classes to be added to the field label',
      control: 'text',
    },
    'label-id': {
      description:
        "A unique `id` to be used on the field label. If one isn't provided, a unique ID will be generated.",
      control: 'text',
    },
    name: {
      description:
        'The `name` is applied to a hidden select element that holds the selected value for the purpose of native HTML form support',
      control: 'text',
    },
    options: {
      description:
        'The list of options to be rendered as an array of objects. If it is defined as an attribute in HTML, it needs to be stringified. Each item must have a `label` and `value`.',
      control: 'object',
    },
    'requirement-label': {
      control: 'text',
      description:
        'Text showing the requirement (ie. "Optional", or "Required").\nIn most cases, this should be used to indicate which fields are optional.\nSee the [form guidelines](https://design.cms.gov/patterns/Forms/forms/) for more info.',
    },
    size: {
      description: 'Sets the max-width of the input either to `"small"` or `"medium"`',
      options: [undefined, 'medium', 'small'],
      control: { type: 'radio' },
    },
    value: {
      description:
        'Sets the selected value. Using this attribute instead of `default-value` means the dropdown will not maintain its own internal selection state.',
      control: 'text',
    },
  },
  args: {
    label: 'Dropdown example',
    name: 'dropdown_field',
    options: JSON.stringify(options),
  },
  parameters: {
    docs: {
      page: WebComponentDocTemplate,
      description: {
        component:
          'For information about how and when to use this component, [refer to its full documentation page](https://design.cms.gov/components/dropdown/).',
      },
      componentEvents: {
        'ds-change': {
          description: 'Dispatched whenever the selected value changes.',
          eventObjectDescription:
            '`event.details.target.value` - The `value` of the selected option',
        },
        'ds-blur': {
          description: 'Dispatched whenever the dropdown loses focus.',
        },
      },
      sharedAttrLists: ['form'],
    },
  },
  decorators: [webComponentDecorator],
};

const Template = (args) => {
  useEffect(() => {
    const onChange = (event) => {
      action('ds-change')(event);
    };
    const onBlur = (event) => {
      action('ds-blur')(event);
    };
    const dropdown = document.querySelector('ds-dropdown');
    dropdown.addEventListener('ds-change', onChange);
    dropdown.addEventListener('ds-blur', onBlur);
    return () => {
      dropdown.removeEventListener('ds-change', onChange);
      dropdown.removeEventListener('ds-blur', onBlur);
    };
  });

  return <ds-dropdown {...args} />;
};

export const Default = Template.bind({});

const htmlOptions = (
  <>
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
    <option value="3">Option 3</option>
    <option value="4">Option 4</option>
    <option value="5">Option 5</option>
    <option value="6">Option 6</option>
    <option value="7">Option 7</option>
    <option value="8" data-extra-attribute="something">
      Option 8
    </option>
  </>
);

export const HtmlOptions = {
  render: Template,
  args: {
    options: undefined,
    label: 'Select an option.',
    hint: 'In this example, options defined as HTML.',
    children: htmlOptions,
  },
};

const htmlOptGroups = (
  <>
    <option value="">- Select an option -</option>
    <optgroup label="Group A">
      <option value="a-1">Option A-1</option>
      <option value="a-2">Option A-2</option>
      <option value="a-3">Option A-3</option>
    </optgroup>
    <optgroup label="Group B" data-extra-attribute="something">
      <option value="b-1">Option B-1</option>
      <option value="b-2">Option B-2</option>
      <option value="b-3">Option B-3</option>
    </optgroup>
  </>
);

export const HtmlOptionGroups = {
  render: Template,
  args: {
    options: undefined,
    label: 'Select an option.',
    hint: 'In this example, options and groups are defined as HTML.',
    children: htmlOptGroups,
  },
};
