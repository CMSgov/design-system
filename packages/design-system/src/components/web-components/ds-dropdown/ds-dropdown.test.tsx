import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import './ds-dropdown';

const defaultAttrs = {
  label: 'Dropdown label',
  options: JSON.stringify([
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2' },
    { label: 'Option 3', value: '3' },
  ]),
};

function renderDropdown(attrs = {}) {
  return render(<ds-dropdown {...defaultAttrs} {...attrs} />);
}

describe('Button', () => {
  it('renders a dropdown', () => {
    const { asFragment } = renderDropdown();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a dropdown using html options', () => {
    const { asFragment } = render(
      <ds-dropdown label={defaultAttrs.label}>
        <>
          <optgroup label="Group A">
            <option value="a-1">Option A-1</option>
            <option value="a-2">Option A-2</option>
          </optgroup>
          <optgroup label="Group B">
            <option value="b-1">Option B-1</option>
            <option value="b-2">Option B-2</option>
          </optgroup>
        </>
      </ds-dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders disabled dropdown', () => {
    renderDropdown({ disabled: 'true' });
    const button = screen.getByRole('button');
    expect(button).toHaveAttribute('disabled');
  });

  it('does not render disabled when passed false', () => {
    renderDropdown({ disabled: 'false' });
    const button = screen.getByRole('button');
    expect(button).not.toHaveAttribute('disabled');
  });

  it('applies additional classes to the wrapper', () => {
    const { container } = renderDropdown({ 'class-name': 'foobar' });
    const wrapper = container.querySelector('.ds-c-dropdown');
    expect(wrapper).toHaveClass('foobar');
  });

  it('applies additional classes to the dropdown button', () => {
    renderDropdown({ 'field-class-name': 'foobar' });
    const button = screen.getByRole('button');
    expect(button).toHaveClass('foobar');
  });

  it('applies size classes', () => {
    renderDropdown({ size: 'small' });
    const button = screen.getByRole('button');
    expect(button.classList.contains('ds-c-field--small')).toBe(true);
  });

  it('fires a custom ds-change event', () => {
    renderDropdown();

    const dropdownRoot = document.querySelector('ds-dropdown');
    const mockHandler = jest.fn();
    dropdownRoot.addEventListener('ds-change', mockHandler);

    const button = screen.getByRole('button');
    userEvent.click(button);
    userEvent.keyboard('{arrowdown}');
    userEvent.keyboard('{enter}');

    expect(mockHandler).toHaveBeenCalledTimes(1);
    dropdownRoot.removeEventListener('ds-change', mockHandler);
  });

  it('fires a custom ds-blur event', async () => {
    renderDropdown();

    const dropdownRoot = document.querySelector('ds-dropdown');
    const onBlur = jest.fn();
    const onChange = jest.fn();
    dropdownRoot.addEventListener('ds-blur', onBlur);
    dropdownRoot.addEventListener('ds-change', onChange);

    const button = screen.getByRole('button');
    userEvent.click(button);
    userEvent.tab();

    const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
    await act(async () => {
      await sleep(40);
    });

    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onChange).not.toHaveBeenCalled();
    dropdownRoot.removeEventListener('ds-blur', onBlur);
    dropdownRoot.removeEventListener('ds-change', onChange);
  });
});
