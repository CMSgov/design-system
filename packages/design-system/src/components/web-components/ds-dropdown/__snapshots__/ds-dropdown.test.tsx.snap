// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders a dropdown 1`] = `
<DocumentFragment>
  <ds-dropdown
    label="Dropdown label"
    options="[{\\"label\\":\\"Option 1\\",\\"value\\":\\"1\\"},{\\"label\\":\\"Option 2\\",\\"value\\":\\"2\\"},{\\"label\\":\\"Option 3\\",\\"value\\":\\"3\\"}]"
  >
    <div
      class="ds-c-dropdown"
    >
      <div
        class="ds-c-label ds-c-dropdown__label"
        id="dropdown--1__label"
      >
        Dropdown label
      </div>
      <p
        aria-atomic="true"
        aria-live="assertive"
        class="ds-c-inline-error"
        id="dropdown--1__error"
      />
      <div
        aria-hidden="true"
        data-a11y-ignore="aria-hidden-focus"
        data-testid="hidden-select-container"
        style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
      >
        <input
          style="font-size: 16px;"
          tabindex="-1"
          type="text"
        />
        <label>
          Dropdown label
          <select
            size="3"
            tabindex="-1"
          >
            <option />
            <option
              value="1"
            >
              Option 1
            </option>
            <option
              value="2"
            >
              Option 2
            </option>
            <option
              value="3"
            >
              Option 3
            </option>
          </select>
        </label>
      </div>
      <button
        aria-controls="dropdown--1__menu"
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-invalid="false"
        aria-labelledby="dropdown--1__button-content dropdown--1__label"
        class="ds-c-dropdown__button ds-c-field"
        id="dropdown--1"
        type="button"
      >
        <span
          class="ds-c-dropdown__label-text"
          id="dropdown--1__button-content"
        >
          Option 1
        </span>
        <span
          class="ds-c-dropdown__caret"
        >
          <svg
            aria-hidden="true"
            class="ds-c-icon ds-u-font-size--sm"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M212.7 148.7c6.2-6.2 16.4-6.2 22.6 0l160 160c6.2 6.2 6.2 16.4 0 22.6s-16.4 6.2-22.6 0L224 182.6 75.3 331.3c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l160-160z"
            />
          </svg>
        </span>
      </button>
    </div>
    <template />
  </ds-dropdown>
</DocumentFragment>
`;

exports[`Dropdown renders a dropdown using html options 1`] = `
<DocumentFragment>
  <ds-dropdown
    label="Dropdown label"
  >
    <div
      class="ds-c-dropdown"
    >
      <div
        class="ds-c-label ds-c-dropdown__label"
        id="dropdown--5__label"
      >
        Dropdown label
      </div>
      <p
        aria-atomic="true"
        aria-live="assertive"
        class="ds-c-inline-error"
        id="dropdown--5__error"
      />
      <div
        aria-hidden="true"
        data-a11y-ignore="aria-hidden-focus"
        data-testid="hidden-select-container"
        style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
      >
        <input
          style="font-size: 16px;"
          tabindex="-1"
          type="text"
        />
        <label>
          Dropdown label
          <select
            size="6"
            tabindex="-1"
          >
            <option />
            <option
              value="a-1"
            >
              Option A-1
            </option>
            <option
              value="a-2"
            >
              Option A-2
            </option>
            <option
              value="b-1"
            >
              Option B-1
            </option>
            <option
              value="b-2"
            >
              Option B-2
            </option>
          </select>
        </label>
      </div>
      <button
        aria-controls="dropdown--5__menu"
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-invalid="false"
        aria-labelledby="dropdown--5__button-content dropdown--5__label"
        class="ds-c-dropdown__button ds-c-field"
        id="dropdown--5"
        type="button"
      >
        <span
          class="ds-c-dropdown__label-text"
          id="dropdown--5__button-content"
        >
          Option A-1
        </span>
        <span
          class="ds-c-dropdown__caret"
        >
          <svg
            aria-hidden="true"
            class="ds-c-icon ds-u-font-size--sm"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M212.7 148.7c6.2-6.2 16.4-6.2 22.6 0l160 160c6.2 6.2 6.2 16.4 0 22.6s-16.4 6.2-22.6 0L224 182.6 75.3 331.3c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l160-160z"
            />
          </svg>
        </span>
      </button>
    </div>
    <template />
  </ds-dropdown>
</DocumentFragment>
`;
